docs/plugins/schedulers.rst

```diff
@@ -1,46 +1,791 @@
+- You may want to consider including host/system information in the id,
+  so your plugin can know when it's running in a place that can actually
+  reference that id. For instance, the raw scheduler starts a local process,
+  but can't very well check the status of a process from a different machine.
+
+An Example
+^^^^^^^^^^
+
+.. code-block:: python
+
+    def _schedule(self, test, kickoff_path):
+        """Submit the kick off script using sbatch.
+
+        :param TestRun test: The TestRun we're kicking off.
+        :param Path kickoff_path: The kickoff script path.
+        :returns: The job id under this scheduler.
+        """
+
+        # We're going to save the slurm log in the test run directory, so it
+        # isn't put just anywhere.
+        slurm_out = test.path/'slurm.log'
+
+        # Run the command to scheduler our batch script.
+        # The default scripts use 'exec >' redirection to redirect all output
+        # script to the kickoff log.
+        # This should be a command that returns when our kickoff script is
+        # in the scheduler queue.
+        proc = subprocess.Popen(['sbatch',
+                                 '--output={}'.format(slurm_out),
+                                 kickoff_path.as_posix()],
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+
+        # Slurm prints the job id when starting an sbatch script, which we
+        # capture and extract.
+        stdout, stderr = proc.communicate()
+
+        # Raise an error if the kickoff was a failure.
+        if proc.poll() != 0:
+            raise SchedulerPluginError(
+                "Sbatch failed for kickoff script '{}': {}"
+                .format(kickoff_path, stderr.decode('utf8'))
+            )
+
+        # Parse out the job id and return it. It will get attached to the
+        # test run object and tracked that way.
+        return stdout.decode('UTF-8').strip().split()[-1]
+
+Cancelling Runs
+~~~~~~~~~~~~~~~
+
+To handle cancelling jobs, we'll be overriding the ``_cancel_job()``
+method of your scheduler class.
+
+You'll need to do the following:
+
+1. (Typically) Compose and run a command to cancel the job given the
+   job id you recorded.
+2. (If cancelling is successful) set 'test.set_run_complete()' to
+   mark the test as complete.
+3. Set the test status to 'STATES.SCHED_CANCELLED'.
+4. Return a ``StatusInfo`` object with the new status of the test, and
+   a reasonable status message.
+
+Additionally, there are four basic cases that need to be handled:
+
+1. The job was never started. This is handled for you in ``cancel_job()``,
+   which calls ``_cancel_job()``.
+2. The job is enqueued but not yet running (or somewhere in between).
+3. The job is running.
+4. The job has finished.
+
+Most of the time, this simply means you will try to cancel the job id,
+and capture any errors.
+
+Additionally, if your job id encodes information that could denote that
```

Comment:
Might be a good place to reference a concept that we've allegedly been looking into of having a central place to submit jobs to any machine.  Could involve having more information for your `scancel` call.

```
-bash-4.2$ scancel --help
Usage: scancel [OPTIONS] [job_id[_array_id][.step_id]]
  -A, --account=account           act only on jobs charging this account
  -b, --batch                     signal batch shell for specified job
  -f, --full                      signal batch shell and all steps for specified job
  -H, --hurry                     avoid burst buffer stage out
  -i, --interactive               require response from user for each job
  -M, --clusters                  clusters to issue commands to.
                                  NOTE: SlurmDBD must be up.
  -n, --name=job_name             act only on jobs with this name
  -p, --partition=partition       act only on jobs in this partition
  -Q, --quiet                     disable warnings
  -q, --qos=qos                   act only on jobs with this quality of service
  -R, --reservation=reservation   act only on jobs with this reservation
      --sibling=cluster_name      remove an active sibling job from a federated job
  -s, --signal=name | integer     signal to send to job, default is SIGKILL
  -t, --state=states              act only on jobs in this state.  Valid job
                                  states are PENDING, RUNNING and SUSPENDED
  -u, --user=user_name            act only on jobs of this user
  -V, --version                   output version information and exit
  -v, --verbose                   verbosity level
  -w, --nodelist                  act only on jobs on these nodes
      --wckey=wckey               act only on jobs with this workload
                                  charactization key

Help options:
  --help                          show this help message
  --usage                         display brief usage message
```